### PR TITLE
Allocate IP addresses for pods running on Kubernetes Windows agents

### PR DESCRIPTION
**Pull request by ofiliz**
This PR allocates IP addresses for Kubernetes Windows agents, bringing Windows to parity with Linux.

- [x] Fix whitespace issues in winagent.t. (This is in its own commit. When reviewing, ignore this commit and review the second commit for a clean diff).
- [x] Allocate IP addresses for containers running on Windows agents.

I still need to install CNI plugin on Windows agents. That will be part of another PR due to external dependencies.

---
This change is [Reviewable](https://reviewable.io/reviews/azure/acs-engine/435)

### CODE REVIEW

File: parts/kuberneteswinagentresourcesvmas.t [hunk 1]
```diff
@@ -1,163 +1,162 @@
...
+      "apiVersion": "[variables('apiVersionDefault')]",
+      "copy": {
+        "count": "[sub(variables('{{.Name}}Count'), variables('{{.Name}}Offset'))]",
+        "name": "vmLoopNode"
+      },
+      "dependsOn": [
+        "[concat('Microsoft.Compute/virtualMachines/', variables('{{.Name}}VMNamePrefix'), copyIndex(variables('{{.Name}}Offset')))]"
+      ],
+      "location": "[variables('location')]",
+      "type": "Microsoft.Compute/virtualMachines/extensions",
```
Comment (JiangtianLi) [inline parts/kuberneteswinagentresourcesvmas.t]: Why is extensions type moved here?

File: parts/kuberneteswindowssetup.ps1 [hunk 2]
```diff
@@ -59,14 +50,37 @@ $global:SecurityGroupName = "{{{nsgName}}}"
...
+$global:AzureVNetPluginURL = "https://github.com/ofiliz/hello/releases/download/pre-release/testw_v0.6.zip"
 
 filter Timestamp {"$(Get-Date -Format o): $_"}
 
 function
 Write-Log($message)
 {
     $msg = $message | Timestamp
-    Write-Output $msg
+    Write-Host $msg
```
Comment (JiangtianLi) [inline parts/kuberneteswindowssetup.ps1]: Why changed to Write-Host? Just for debugging? I think we want to redirect output to log file.

File: examples/networkpolicy/kubernetes-azure-hybrid.json [hunk 3]
```diff
@@ -18,16 +18,14 @@
         "count": 2,
         "vmSize": "Standard_D2_v2",
         "availabilityProfile": "AvailabilitySet",
-        "ipAddressCount": 20,
         "osType": "Linux"
       },
       {
-        "name": "linuxpool2",
+        "name": "windowspool1",
```
Comment (JiangtianLi) [inline examples/networkpolicy/kubernetes-azure-hybrid.json]: Please make sure to execute make ci to run the checkin validation tests.

File: parts/kuberneteswinagentresourcesvmas.t [hunk 4]
```diff
@@ -1,163 +1,171 @@
...
-        "typeHandlerVersion": "1.8",
-        "autoUpgradeMinorVersion": true,
-        "protectedSettings": {
-          "commandToExecute": "[concat('powershell.exe -ExecutionPolicy Unrestricted -command \"', '$arguments = ', variables('singleQuote'),'-MasterIP ',variables('masterInternalLbIp'),' -KubeDnsServiceIp ',variables('kubeDnsServiceIp'),' -MasterFQDNPrefix ',variables('masterFqdnPrefix'),' -Location ',variables('location'),' -AgentKey ',variables('clientPrivateKey'),' -AzureHostname ',variables('{{.Name}}VMNamePrefix'),copyIndex(variables('{{.Name}}Offset')),' -AADClientId ',variables('servicePrincipalClientId'),' -AADClientSecret ',variables('servicePrincipalClientSecret'),variables('singleQuote'), ' ; ', variables('windowsCustomScriptSuffix'), '\" > %SYSTEMDRIVE%\\AzureData\\CustomDataSetupScript.log 2>&1')]"
-        }
-      }, 
-      "type": "Microsoft.Compute/virtualMachines/extensions"
-    }
-    
+    {
```
Comment (anhowe) [inline parts/kuberneteswinagentresourcesvmas.t]: why is everything different in the file?  Did you adjust whitespace?

File: parts/kuberneteswindowssetup.ps1 [hunk 5]
```diff
@@ -303,14 +383,11 @@ try
         Write-Log "Create the Pause Container kubletwin/pause"
         New-InfraContainer
 
-        Write-Log "Get the POD CIDR"
-        $podCIDR = Get-PodCIDR
-
-        Write-Log "write kubelet startfile with pod CIDR of $podCIDR"
-        Write-KubernetesStartFiles $podCIDR
+        Write-Log "Configure networking"
+        Set-NetworkConfig
```
Comment (anhowe) [inline parts/kuberneteswindowssetup.ps1]: do you want to do this if they don't use CNI?

File: parts/kuberneteswinagentresourcesvmas.t (hunk 4 above)
Comment (ofiliz) [inline parts/kuberneteswinagentresourcesvmas.t]: Please see the comment in the PR description. Yes, I wanted windows and Linux agentresourcesvmas.t files to have a clean diff so I can see what is missing from Windows. There were too many CRLF, trailing spaces, tab/space issues. I fixed the whitespace issues as a separate commit. Review the second commit.

File: parts/kuberneteswinagentresourcesvmas.t (hunk 1 above)
Comment (ofiliz) [inline parts/kuberneteswinagentresourcesvmas.t]: To match its place in linuxagentresourcesvmas.t. I want a clean diff between these two files to make sure the templates are same as much as possible.

File: parts/kuberneteswindowssetup.ps1 (hunk 2 above)
Comment (ofiliz) [inline parts/kuberneteswindowssetup.ps1]: Write-Host also outputs to the log file. The difference is Write-Output also acts as a "return" statement, making it hard to log in functions that also return a value.

This isn't actually necessary for the purposes of this PR. I just thought it is a good improvement.

File: examples/networkpolicy/kubernetes-azure-hybrid.json (hunk 3 above)
Comment (ofiliz) [inline examples/networkpolicy/kubernetes-azure-hybrid.json]: Sure, will do.

**Comment by ofiliz** [comment]
Any more comments?

File: parts/kuberneteswinagentresourcesvmas.t (hunk 1 above)
Comment (JiangtianLi) [inline parts/kuberneteswinagentresourcesvmas.t]: Sorry for the delay. Just curious, where is linuxagentresourcesvmas.t?

Comment (ofiliz) [inline parts/kuberneteswinagentresourcesvmas.t]: The actual file name is kubernetesagentresourcevmas.t. It is the Linux equivalent of this file (kuberneteswinagentresourcevmas.t).

**Comment by ofiliz** [comment]
This PR has been waiting for more than a month. Do you have any other comments? If not, I'd like to rebase this again, hopefully for the last time.

**Comment by JiangtianLi** [comment]
@anhowe @colemickens Can you take a look?

**Comment by anhowe** [comment]
Does the outbound internet work with this solution?

**Comment by anhowe** [comment]
@ofiliz I would like to understand better the fix or plans to fix around outbound internet #561

**Comment by ofiliz** [comment]
This PR has nothing to do with pod outbound internet connectivity. Completely separate. Let's not gate one on the other.

**Comment by anhowe** [comment]
@ofiliz - if this fix goes in I'm worried we will get support questions on "why doesn't outbound internet not work with CNI for windows"?  Currently we say, CNI is not supported for Windows.  Let me ask around to see if it will be ok to say Outbound internet for CNI is not supported for Windows.

**Comment by shrutir25** [comment]
@anhowe - how do we want to proceed on this issue

**Comment by seanknox** [comment]
I'm not comfortable merging in such a substantial change without test coverage. Perhaps we can have someone from Ex-Deis pick this branch up/rebase/add tests to free Onur up. Will defer to you, @anhowe.

**Comment by ofiliz** [comment]
Substantial change? It is 5 lines? Please read the description.

**Comment by seanknox** [comment]
> Substantial change? It is 5 lines? Please read the description.

It's nearly 200 lines of changes, according to the diff.

**Comment by bacongobbler** [comment]
I can see that the actual code that was changed within the template is only 5 lines, but I can also see @seanknox's point about the file's whitespacing was changed in this PR. Is that necessary or can that be pulled out separately?

**Comment by bacongobbler** [comment]
I missed [this comment](https://github.com/Azure/acs-engine/pull/435/commits/d3ffcd0931dbf2feac565d42152237b806b70b42#r108581875) explaining the rationale. Don't have any handle on how acs-engine works but from a Go template point of view, this looks good to me. Don't consider my naive experience as an official LGTM, though.

**Comment by dmitsh** [comment]
@acs-bot test this please

**Comment by dmitsh** [comment]
@acs-bot test this please

**Pull request merged**